### ISSUE TITLE
Heap buffer overflow in print_raw_cipherlist (apps/s_cb.c)

### ISSUE THREAD

**Issue by bernd-edlinger** [openssl/openssl#2189]
I built using recent gcc and OpenSSL-1_1_0-stable
./config -g -fsanitize=address -fno-omit-frame-pointer
make
make test V=1 2>&1 | tee test.log

in test.log I got

```
depth=0 C = UK, O = OpenSSL Group, OU = FOR TESTING PURPOSES ONLY, CN = Test Server Cert
verify error:num=21:unable to verify the first certificate
verify return:1
CONNECTION ESTABLISHED
Protocol version: TLSv1.2
Client cipher list: 0x0000=================================================================
==25110==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6020000004f3 at pc 0x2b15e44b2391 bp 0x7ffe30b7cc00 sp 0x7ffe30b7cbf8
READ of size 1 at 0x6020000004f3 thread T0
DONE
    #0 0x2b15e44b2390 in ssl3_get_cipher_by_char ssl/s3_lib.c:3550
    #1 0x4757c7 in print_raw_cipherlist apps/s_cb.c:975
    #2 0x4757c7 in print_ssl_summary apps/s_cb.c:1073
    #3 0x482caf in rev_body apps/s_server.c:3089
    #4 0x48a474 in do_server apps/s_socket.c:179
    #5 0x488547 in s_server_main apps/s_server.c:1934
    #6 0x41ef90 in do_cmd apps/openssl.c:471
    #7 0x4202fc in main apps/openssl.c:177
    #8 0x2b15e5351f44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)
    #9 0x4209f6  (/home/ed/OPC/openssl/apps/openssl+0x4209f6)

0x6020000004f3 is located 0 bytes to the right of 3-byte region [0x6020000004f0,0x6020000004f3)
allocated by thread T0 here:
    #0 0x2b15e3585ef0 in __interceptor_malloc ../../../../gcc-trunk/libsanitizer/asan/asan_malloc_linux.cc:62
    #1 0x2b15e4a8e687 in CRYPTO_memdup crypto/o_str.c:67
    #2 0x2b15e4507f83 in PACKET_memdup ssl/statem/../packet_locl.h:385
    #3 0x2b15e4507f83 in ssl_bytes_to_cipher_list ssl/statem/statem_srvr.c:3271
    #4 0x2b15e4507f83 in tls_process_client_hello ssl/statem/statem_srvr.c:1175
    #5 0x2b15e44e7133 in read_state_machine ssl/statem/statem.c:589
    #6 0x2b15e44e7133 in state_machine ssl/statem/statem.c:385
    #7 0x2b15e44d4724 in SSL_do_handshake ssl/ssl_lib.c:3025
    #8 0x2b15e4490eca in ssl_ctrl ssl/bio_ssl.c:349
    #9 0x2b15e48f34b8 in buffer_ctrl crypto/bio/bf_buff.c:343
    #10 0x482c7a in rev_body apps/s_server.c:3063
    #11 0x48a474 in do_server apps/s_socket.c:179
    #12 0x488547 in s_server_main apps/s_server.c:1934
    #13 0x41ef90 in do_cmd apps/openssl.c:471
    #14 0x4202fc in main apps/openssl.c:177
    #15 0x2b15e5351f44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)

SUMMARY: AddressSanitizer: heap-buffer-overflow ssl/s3_lib.c:3550 in ssl3_get_cipher_by_char
Shadow bytes around the buggy address:
  0x0c047fff8040: fa fa 00 01 fa fa 00 00 fa fa 03 fa fa fa 00 06
  0x0c047fff8050: fa fa fd fa fa fa fd fd fa fa 00 06 fa fa 00 06
  0x0c047fff8060: fa fa 03 fa fa fa 00 06 fa fa fd fa fa fa fd fd
  0x0c047fff8070: fa fa 00 00 fa fa 00 fa fa fa 03 fa fa fa 05 fa
  0x0c047fff8080: fa fa 00 00 fa fa 00 01 fa fa 00 00 fa fa 00 fa
=>0x0c047fff8090: fa fa fd fd fa fa fd fa fa fa fd fd fa fa[03]fa
  0x0c047fff80a0: fa fa fd fd fa fa fd fa fa fa fd fa fa fa fd fa
  0x0c047fff80b0: fa fa 00 fa fa fa fd fa fa fa fd fa fa fa fd fa
  0x0c047fff80c0: fa fa fd fa fa fa fd fa fa fa fd fa fa fa fd fa
  0x0c047fff80d0: fa fa fd fa fa fa fd fa fa fa fd fa fa fa fd fa
  0x0c047fff80e0: fa fa fd fd fa fa fd fa fa fa fd fa fa fa fd fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==25110==ABORTING
engine "ossltest" set.
engine "ossltest" set.
CONNECTION FAILURE
47964245314304:error:1417D0FC:SSL routines:tls_process_client_hello:unknown protocol:ssl/statem/statem_srvr.c:956:
write:errno=0
```

I don't know how to locate the problematic test case.
Probably near  "../test/recipes/70-test_sslextension.t ....."

The cipher list is 3 bytes long, thus rlistlen=3 is not a multiple of num=2
rlist = [0x00 0x00 0x2F]

Although this is a fatal error (with non-zero return code),
make test prints at the end:

```
All tests successful.
Files=90, Tests=479, 329 wallclock secs ( 2.94 usr  0.50 sys + 186.48 cusr 152.72 csys = 342.64 CPU)
Result: PASS
make[1]: Leaving directory `/home/ed/OPC/openssl'
```

**Comment by kroeckx** [comment]
It's also in the travis logs, but only in the 1.1.0-stable branch.

**Comment by kroeckx** [comment]
So there are several issues here:
- SSL_get0_raw_cipherlist() doesn't seem to be consistent. It now always return TLS_CIPHER_LEN (== 2) even when there was a SSLv2 compatible client hello doing SSLv3+. In SSLv2 the cipher list was 3 bytes but in SSLv3 it's 2 bytes. The raw list that is actually returned has the 3 byte list like send in the client hello. So we either need to revert the commit (and probably some others) that made it always say there are 2 bytes per cipher, or actually return a list that has 2 byte ciphers.
- We don't get this error in the master branch while the code that has the problem is just the same. So we might have lost test coverage somehow.
- The test aborted, but we said everything is ok, which defeats the purpose of actually letting travis check this. I wonder which return code asan generates in that case.

**Comment by bernd-edlinger** [comment]
I only use gcc not clang.
In gcc the application terminates with exit code 1 when that kind of error
happens the server should terminate immediately.
I think the return code is simply thrown away here:
```
    if ($self->serverconnects == 0) {
        die "serverpid is zero\n" if $self->serverpid == 0;
        print "Waiting for server process to close: "
              .$self->serverpid."\n";
        waitpid( $self->serverpid, 0);
    }
```
It would also be 1 if a memory leak happens, although that is only checked when
the application terminates normally.

You can change the exit code with ASAN_OPTIONS="exitcode=123" for instance.

I see the same error happening in master too.
Just use "make test V=1", why are you not seeing this?

**Comment by bernd-edlinger** [comment]
Oh, that is the difference:

1.1.0 branch is built with these options:
export CONFIG_OPTS="no-asm enable-asan enable-rc5 enable-md2"

but master is built with slightly different options:
export CONFIG_OPTS="no-asm enable-asan enable-rc5 enable-md2 no-shared -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION"

the no-shared skips a few test cases...

**Comment by mattcaswell** [comment]
> SSL_get0_raw_cipherlist() doesn't seem to be consistent. It now always return TLS_CIPHER_LEN (== 2) even when there was a SSLv2 compatible client hello doing SSLv3+. In SSLv2 the cipher list was 3 bytes but in SSLv3 it's 2 bytes. The raw list that is actually returned has the 3 byte list like send in the client hello. So we either need to revert the commit (and probably some others) that made it always say there are 2 bytes per cipher, or actually return a list that has 2 byte ciphers.

The commit that caused it to always return 2 is not the source of the problem and reverting it will not solve the issue. Prior to 1.1.0 it asked the SSL_METHOD what the size of its ciphersuite encoding was. The SSLv2 method would return 3, and everything else would return 2. So, even in 1.0.2, if you have negotiated SSLv3+ with an SSLv2 compat ClientHello then SSL_get0_raw_cipherlist() will still return 2!!

The difference in 1.0.2 (that means it doesn't crash) is that version negotiation works completely differently. In 1.0.2 there is an initial stage on the front of the state machine which works out which SSL_METHOD needs to be used. If it receives a pure SSLv2 ClientHello it starts the SSLv2 method and if it receives an SSLv3+ ClientHello it uses an SSLv3+ method. However, if it receives an SSLv2 compat ClientHello but wants to use SSLv3+ then it constructs a "fake" SSLv3+ format ClientHello and then lets the SSLv3+ method handle it. It is the raw ciphersuite list from the "fake" ClientHello that gets stored away, so when you call SSL_get0_raw_cipherlist() in this scenario you're not actually getting the raw list at all - you are getting the "fake" list which has had all the SSLv2 ciphersuites removed from it and it has been converted into SSLv3+ format. Therefore the value of 2 for the length of a cipher is still correct.

In 1.1.0 things works completely differently and we no longer create a "fake" SSLv3+ ClientHello. We process the SSLv2 compat ClientHello directly. That means when you call SSL_get0_raw_cipherlist() you really do get the raw cipherlist in whatever format we received it in.

I think the correct solution here is to pre-process the list to convert it into SSLv3+ format before we save it away.

**Comment by mattcaswell** [comment]
Fix for this in #2280.
